support-frontend/assets/components/asyncronously/asyncronously.ts

```diff
@@ -1,47 +1,34 @@
-import type { ComponentType, Node } from 'react';
-import { Component } from 'react';
+import { useEffect, useState } from 'react';
+import type { ComponentType } from 'react';
 
-type PropTypes = {
+type PropTypes<T> = {
 	loader: Promise<{
-		default: ComponentType<any>;
+		default: ComponentType<T>;
 	}>;
-	loading: Node;
-	children: (arg0: ComponentType<any>) => Node;
+	loading: JSX.Element;
+	render: (Component: ComponentType<T>) => JSX.Element;
 };
-type State = {
-	loaded: boolean;
-};
-
-class Asyncronously extends Component<PropTypes, State> {
-	static defaultProps = {
-		loading: null,
-	};
-	state = {
-		loaded: false,
-	};
-
-	componentDidMount() {
-		const { loader } = this.props;
-		loader.then((imported) => {
-			this.component = imported.default;
-			this.setState({
-				loaded: true,
-			});
-		});
-	}
 
-	component = null;
+function Asyncronously<T>(props: PropTypes<T>): JSX.Element | null {
+	const [component, setComponent] = useState<ComponentType<T>>();
 
-	render() {
-		const { loading, children } = this.props;
-		const { loaded } = this.state;
+	const { loading, render, loader } = props;
 
-		if (loaded && this.component) {
-			return children(this.component);
-		}
+	useEffect(() => {
+		void loader.then((imported) => {
+			setComponent(imported.default);
```

Comment:
I think we'll need to do 

```ts
setComponent(() => imported.default);
```

This is because passing a function to `setComponent` will cause it to call the function and store the return value. So if we don't wrap it, we'll be storing a `JSX.Element`, not the component. 

If we make this change we can then use it like `<Component />` instead of `{Component}`

Reply:
Did you try this out locally? It doesn't work for me

Reply:
I didn't! I only tried it in a fresh `create-react-app` project - I'll have a go locally

Reply:
Okay - it did work but I had to pass in some extra arguments that are currently not being passed. This worked: 

```ts
render={(MktConsent) => (
  <MktConsent requestPending={false} error={false} />
)}
```

Reply:
Are there particular disadvantages you have in mind in regards to returning a JSX.Element vs a component in this instance? It seems to me having to add extra arguments to get this to work may be unnecessary

Reply:
Yeah I think it's probably best to store the component instead of the jsx.

Setters from `useState` can be passed two type of arguments, either: a new state; or a function mapping from old state to new state. So if I have


```ts
const [count, setCount] = useState(0);
```

We can write

```ts
const increment = () => setCount(count + 1);
```

or 

```ts
const increment = () => setCount((c) => c + 1);
```

The second one allows us to remove a dependence on `count` e.g we could do 

```ts
const increment = useCallback(() => setCount((c) => c + 1), []);
```

If we wanted to just call `setCount(count)`, we'd have to add count to the dependency array, which would mean we recreate the `increment` function every time. 

I don't think we should rely on this behaviour to call `imported.default`. It's quite subtle and easy to miss and also not really what it's intended for. I also think this will break if we import a class component instead of a function one. If we import a class component, I think we'll actually be storing the component instead of jsx (this will break if we try and render it like `{MktConsent}` instead of `<MktConsent />`).

I think storing the component also makes it more flexible and reusable e.g if I have a component

```tsx
function Hello({ name }) {
  return <div>Hello, {name}<div />
}
```

Then we could do 

```tsx
render={(Hello) => <Hello name="Maria" />}
```

or

```tsx
render={(Hello) => <Hello name="Tom" />}
```

depending on the context we're using the component.

Reply:
In this instance as the set state function is only called once after the initial render I wouldn’t have thought we’d need to take the callback approach for performance reasons.

Not used imported.default too much myself, would be good to read into its proper usage if you have a reference handy. From what I can tell from this components history, I think this has always been a slightly hacky way to asynchronously render only the marketing consent component on thank you pages but I suppose could be used for other reasons and this possibly breaking if it were to be used with a class component is concerning. 

From one of your earlier messages I think you’d managed to get this working by passing some extra props. I’m not too sure how you’d changed the component to use those props but am more than happy for you to push those changes up to this branch and then we can then rope in another reviewer to have a go over this

Reply:
Happy to merge, I'll sort these conflicts then get this in!